### DISCUSSION TITLE
Shared library based MATLAB wrapper

### DISCUSSION THREAD

**Discussion by stephen-young** [CoolProp/CoolProp#2368] [Show and tell]
The official MATLAB wrapper uses MATLAB's python interface to use the CoolProp python wrapper. In my work for my Masters and PhD work [modelling bubbly flows in MATLAB](https://www.sciencedirect.com/science/article/pii/S0009250922004559), I had written a rough wrapper that used MATLAB's interface with C libraries to call CoolProp's shared library functions. I've done some work recently to tidy it up to a point that I think it's worth sharing. You can view the code here: https://github.com/stephen-young/coolprop-mat

Let me know what you think! In my personal experience, using the DLL interface reduced the friction in sharing my MATLAB code with colleagues so I don't think it would hurt to have it has another option for people.

**Comment by ibell** [comment]
My experience has been that calling shared libraries with MATLAB is a very unpleasant experience, not the least because you need to install a complete compiler stack just to call shared libraries (no other programming language I have worked with has this requirement). It is easy to get python installed, and debugging of the python interface is much easier (usually) than the obtuse MATLAB issues with loading shared libraries. And don't get me started on thunk files...

But perhaps my MATLAB knowledge is outdated and these constraints no longer apply. If so, I would be happy to consider your solution.

**Reply by stephen-young** [reply]
It does still require you to have a C compiler installed on your system but their [MinGW compiler add-on](https://www.mathworks.com/matlabcentral/fileexchange/52848-matlab-support-for-mingw-w64-c-c-fortran-compiler) does make satisfying this requirement fairly easy. In my experience, I have not had any issues loading library. As long as the paths to the DLL and header file are valid, loading the library is not a problem. I have run into edge cases where MATLAB will crash upon calling a DLL function incorrectly (the most recent one I can remember was due to an empty double being passed to a library function that asked for an integer and this some how bypassed MATLAB's function signature and input type checks) but by and large, those kind of issues have been rare.

I don't know what requirements this could impose on the C++ side or what thunk files are but over the past few years, I have been able to use the pre-compiled binaries found on the SourceForge page for the project. I have mainly been evaluating the properties of water and the main constituents of air (nitrogen, oxygen, argon, carbon dioxide). Perhaps there are edge cases out there that my wrapper has not accounted for at the moment. I have done some basic testing to check the functionality but only on Windows.

My intent is not to replace the python interface but to supplement it with this additional option but I understand the reluctance to add something to the maintenance list. I believe I have written the wrapper in a way that would make that maintenance easy and would only need changes if the DLL API changes.

**Reply by ibell** [reply]
Even MinGW is not easy to get working in all cases! I get the appeal of this solution, what about just adding a PR with a link to your solution?

Are thunk files no longer a thing with MATLAB? That was hugely annoying previously.

**Reply by stephen-young** [reply]
Ah, I see. Accessing `gcc` through MinGW has worked for me so far.

After a little internet searching it appears that thunk files could still be a thing but they are a [poorly documented part of MATLAB](https://ofekshilon.com/2016/07/15/on-matlabs-loadlibrary-proto-file-and-pcwin64-thunk/). However, the posts I've seen about them seem to be in 2013 to 2016. MATLAB might generate thunk files under the hood or they might be a vestige from the transition to 64-bit. I first started using the DLL to access the low-level interface from MATLAB in around 2018 and I've only ever had to provide a path to the DLL and the header file for the DLL. 

For the PR, are you just looking for a link and some explanatory text in the [MATLAB wrapper README](https://github.com/CoolProp/CoolProp/blob/master/wrappers/MATLAB/README.rst)?

**Reply by ibell** [reply]
Yup, that sounds about right
